ci/Jenkinsfile_utils.groovy

```diff
@@ -18,18 +18,23 @@
 // under the License.
 
 // initialize source codes
-def init_git() {
+// Clean - 0: None, 1: xdf, 2: xdff
+def init_git(clean=2) {
   deleteDir()
   retry(5) {
     try {
       // Make sure wait long enough for api.github.com request quota. Important: Don't increase the amount of
       // retries as this will increase the amount of requests and worsen the throttling
       timeout(time: 15, unit: 'MINUTES') {
         checkout scm
-        sh 'git clean -xdff'
+        if(clean == 1)
+          clean = 'git clean -xdf'
+        else if(clean == 2)
+          clean = 'git clean -xdff'
```

Comment:
Is there a difference between the ```clean``` command in clean == 1 and clean == 2 ?

Reply:
Please leave this as it is.

The workspace always has to be considered empty and reproducible. If you need to persist something in between stages, use the stashing mechanism. With this change you are circumventing a problem - that you are not persisting state properly - and it will blow up as soon as we have more than one restricted slave and job one gets scheduled onto a different slave than job two

Reply:
This change is not to persist something in between stages. During maven deploy, it attempts to move all code into a temp directory, rebuild, and deploy from that temp directory. The rebuilding calls "git submodule update" which breaks if the .git directory has already been cleaned with "-ff" since the submodule configuration is deleted. I keep the default as "clean -xdff" so that keeping the .git directory must be explicitly chosen and does not affect any other code.

Reply:
I think I'd like to have a deeper look into that, so far I'm not following

Reply:
The maven release plugin messes with the project SCM (git). It first tries to create a new tag for the release and a new commit updating the version number located in the POM file. Then, when performing the release, it will checkout the code to a temporary directory and perform the release (building, testing, and deploying) from that new clean directory. This will break if the .git directory has been pruned. You can see docs for the maven release plugin at http://maven.apache.org/maven-release/maven-release-plugin/examples/prepare-release.html